RMtools_3D/do_fitIcube.py

```diff
@@ -70,77 +77,122 @@ def main():
+    freqArr_Hz = get_freq_array (I_filename) 
+       
+    # Run polynomial fitting on the spectra
+    make_model_I(datacube     = datacube, 
+                 header       = headI, 
+                 freqArr_Hz   = freqArr_Hz,
                  polyOrd      = args.polyOrd,
                  cutoff       = args.cutoff,
                  prefixOut    = args.prefixOut,
                  outDir       = dataDir,
+                 nBits        = 32,
+                 threshold    = args.threshold,
+                 apply_mask   = args.apply_mask,
+                 num_cores    = args.num_cores,
                  verbose        = args.verbose,
-                 fit_function = args.fit_function)
+                 fit_function = args.fit_function, 
+                 save_mask    = args.save_mask, 
+                 save_coefficients = args.save_coeff,
+                 save_modelIcube = args.save_model_Icube)
 
 
 #-----------------------------------------------------------------------------#
-def make_model_I(fitsI, freqFile, polyOrd=2, cutoff=-1, prefixOut="",
-                 outDir="", verbose=True,fit_function='log'):
-    """
-    Detect emission in a cube and fit a polynomial or power law model spectrum 
-    to the emitting pixels. Create a representative noise spectrum using the 
-    residual planes.
-    """
+
+
+def open_datacube(fitsI, verbose=True):
 
     # Default data type
     
     # Sanity check on header dimensions
     print("Reading FITS cube header from '%s':" % fitsI)
-    headI,datacube=readFitsCube(fitsI, verbose)
-    
+    header, datacube = readFitsCube(fitsI, verbose)
     
     nDim = datacube.ndim
     if nDim < 3 or nDim > 4:
         print("Err: only 3 or 4 dimensions supported: D = %d." % headI["NAXIS"])
         sys.exit()
 
-
     # freq_axis=find_freq_axis(headI) 
     # #If the frequency axis isn't the last one, rotate the array until it is.
     # #Recall that pyfits reverses the axis ordering, so we want frequency on
     # #axis 0 of the numpy array.
     # if freq_axis != 0 and freq_axis != nDim:
     #     datacube=np.moveaxis(datacube,nDim-freq_axis,0)
+    
+    return datacube, header
+    
+    
+def get_frequencies(datacube, headI, freqFile):
 
-
-    nBits=np.abs(headI['BITPIX'])    
+    nBits = np.abs(headI['BITPIX'])    
     dtFloat = "float" + str(nBits)
 
-    
-    nChan = datacube.shape[0]
+    nChan = datacube.shape[0] # for now, assumes frequency is the first axis
     
     # Read the frequency vector
     print("Reading frequency vector from '%s'." % freqFile)
     freqArr_Hz = np.loadtxt(freqFile, dtype=dtFloat)
+    if nChan!=len(freqArr_Hz):
+        print("Err: frequency vector and frequency axis of cube unequal length.")
+        sys.exit()
+        
+    return freqArr_Hz
+   
+   
+def cube_noise(datacube, header, freqArr_Hz, cutoff=-1, threshold=3):
+
+
+    """
+    Estimate channel noise of a cube data. Returns rms values and a mask 2D data.
+     
```

Comment:
It would be nice to have a clearly explanation of how the noise is estimated, and the exact meanings of the two parameters (cutout and threshold). I'm currently very confused about the distinction between the two.

Reply:
I have removed cutoff. Only threshold. Here is the definition: